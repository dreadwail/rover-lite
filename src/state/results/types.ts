export interface Badge {
  readonly content: string;
  readonly imageUrl: string;
  readonly title: string;
}

export interface Listing {
  readonly thumbnailUrl: string;
  readonly rank: number;
  readonly name: string;
  readonly badges: Badge[];
  readonly tagline: string;
  readonly neighborhood: string;
  readonly city: string;
  readonly state: string;
  readonly zip: string;
  readonly repeatClientCount: number;
  readonly reviewCount: number;
  readonly ratingAverage: number;
  readonly featuredReview: string;
  readonly currencySymbol: string;
  readonly price: number;
  readonly priceUnit: string;
  readonly webUrl: string;
}

export interface Results {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly isStale: boolean;
  readonly listings: Listing[];
}
