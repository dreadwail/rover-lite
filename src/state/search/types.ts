import { SearchType } from '../searchTypes/types';

export type SearchInputs = Record<string, any>;

export interface Search {
  readonly searchType: SearchType;
  readonly latitude: number;
  readonly longitude: number;
  readonly inputs: SearchInputs;
}
