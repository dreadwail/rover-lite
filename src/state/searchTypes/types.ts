export type SearchType =
  | 'overnight-boarding'
  | 'overnight-traveling'
  | 'drop-in'
  | 'doggy-day-care'
  | 'dog-walking';

export type Endpoints = { [T in SearchType]?: string };

export interface SearchTypes {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly endpoints: Endpoints;
}
