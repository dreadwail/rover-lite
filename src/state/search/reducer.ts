import { Reducer } from 'redux';

import { SearchFieldChangedAction, SearchTypeChangedAction } from '../../domains/search/actions';

import { Search } from './types';

export type SearchAction = SearchTypeChangedAction | SearchFieldChangedAction;

const hardCodedLatitude = 47.6843725;
const hardCodedLongitude = -122.3874837;

const initialState: Search = {
  searchType: 'overnight-boarding',
  latitude: hardCodedLatitude,
  longitude: hardCodedLongitude,
  inputs: {},
};

const reducer: Reducer<Search, SearchAction> = (state = initialState, action) => {
  switch (action.type) {
    case 'SEARCH.TYPE_CHANGED':
      return { ...initialState, searchType: action.payload };
    case 'SEARCH.FIELD_CHANGED':
      return {
        ...state,
        inputs: {
          ...state.inputs,
          [action.payload.key]: action.payload.value,
        },
      };
    default: {
      return state;
    }
  }
};

export default reducer;
