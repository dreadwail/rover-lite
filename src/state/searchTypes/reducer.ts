import { Reducer } from 'redux';

import {
  FetchSearchTypesFulfilledAction,
  FetchSearchTypesPendingAction,
  FetchSearchTypesRejectedAction,
} from '../../domains/searchTypes/actions';

import { SearchTypes } from './types';

export type FetchSearchTypesAction =
  | FetchSearchTypesPendingAction
  | FetchSearchTypesFulfilledAction
  | FetchSearchTypesRejectedAction;

const initialState: SearchTypes = {
  isFetching: false,
  hasFailed: false,
  endpoints: {},
};

const reducer: Reducer<SearchTypes, FetchSearchTypesAction> = (state = initialState, action) => {
  switch (action.type) {
    case 'FETCH_SEARCH_TYPES.PENDING':
      return { ...state, isFetching: true, hasFailed: false };
    case 'FETCH_SEARCH_TYPES.FULFILLED':
      return { ...state, isFetching: false, hasFailed: false, endpoints: action.payload };
    case 'FETCH_SEARCH_TYPES.REJECTED':
      return { ...state, isFetching: false, hasFailed: true };
    default: {
      return state;
    }
  }
};

export default reducer;
