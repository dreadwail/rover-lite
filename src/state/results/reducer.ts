import { Reducer } from 'redux';

import {
  FetchResultsFulfilledAction,
  FetchResultsPendingAction,
  FetchResultsRejectedAction,
} from '../../domains/results/actions';
import { SearchFieldChangedAction, SearchTypeChangedAction } from '../../domains/search/actions';

import { Results } from './types';

export type ResultsAction =
  | FetchResultsPendingAction
  | FetchResultsFulfilledAction
  | FetchResultsRejectedAction
  | SearchTypeChangedAction
  | SearchFieldChangedAction;

const initialState: Results = {
  isFetching: false,
  hasFailed: false,
  isStale: true,
  listings: [],
};

const reducer: Reducer<Results, ResultsAction> = (state = initialState, action) => {
  switch (action.type) {
    case 'FETCH_RESULTS.PENDING':
      return { ...state, isFetching: true, hasFailed: false, isStale: false };
    case 'FETCH_RESULTS.FULFILLED':
      return {
        ...state,
        isFetching: false,
        hasFailed: false,
        isStale: false,
        listings: action.payload,
      };
    case 'FETCH_RESULTS.REJECTED':
      return { ...state, isFetching: false, hasFailed: true };
    case 'SEARCH.TYPE_CHANGED':
      return { ...state, isStale: true };
    case 'SEARCH.FIELD_CHANGED':
      return { ...state, isStale: true };
    default: {
      return state;
    }
  }
};

export default reducer;
