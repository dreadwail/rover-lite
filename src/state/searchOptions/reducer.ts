import { Reducer } from 'redux';

import { SearchTypeChangedAction } from '../../domains/search/actions';
import {
  FetchSearchOptionsFulfilledAction,
  FetchSearchOptionsPendingAction,
  FetchSearchOptionsRejectedAction,
} from '../../domains/searchOptions/actions';

import { SearchOptions } from './types';

export type FetchSearchOptionsAction =
  | FetchSearchOptionsPendingAction
  | FetchSearchOptionsFulfilledAction
  | FetchSearchOptionsRejectedAction
  | SearchTypeChangedAction;

const initialState: SearchOptions = {
  isFetching: false,
  hasFailed: false,
  options: {},
};

const reducer: Reducer<SearchOptions, FetchSearchOptionsAction> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case 'FETCH_SEARCH_OPTIONS.PENDING':
      return { ...state, isFetching: true, hasFailed: false };
    case 'FETCH_SEARCH_OPTIONS.FULFILLED':
      return { ...state, isFetching: false, hasFailed: false, options: action.payload.options };
    case 'FETCH_SEARCH_OPTIONS.REJECTED':
      return { ...state, isFetching: false, hasFailed: true };
    case 'SEARCH.TYPE_CHANGED':
      return initialState;
    default: {
      return state;
    }
  }
};

export default reducer;
