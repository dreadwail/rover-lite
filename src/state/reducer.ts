import { combineReducers } from 'redux';

import results from './results/reducer';
import search from './search/reducer';
import searchOptions from './searchOptions/reducer';
import searchTypes from './searchTypes/reducer';
import { RoverSearchState } from './types';

export default () =>
  combineReducers<RoverSearchState>({
    searchTypes,
    searchOptions,
    search,
    results,
  });
