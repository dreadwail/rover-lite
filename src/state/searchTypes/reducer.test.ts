import reducer, { FetchSearchTypesAction } from './reducer';
import { SearchTypes } from './types';

const unhandledAction = ({ type: 'UNHANDLED_TEST_ACTION' } as any) as FetchSearchTypesAction;

describe('search types reducer', () => {
  let oldState: SearchTypes;
  let newState: SearchTypes;

  beforeEach(() => {
    oldState = {
      isFetching: false,
      hasFailed: false,
      endpoints: {},
    };
  });

  describe('when passed no previous state', () => {
    beforeEach(() => {
      oldState = undefined as any;
      newState = reducer(undefined, unhandledAction);
    });

    it('returns the correct initial state', () => {
      expect(newState).toMatchSnapshot();
    });
  });

  describe('when receiving an unhandled action', () => {
    beforeEach(() => {
      newState = reducer(oldState, unhandledAction);
    });

    it('returns the old state', () => {
      expect(newState).toEqual(oldState);
    });
  });

  describe('when fetching search types starts', () => {
    beforeEach(() => {
      const action: FetchSearchTypesAction = { type: 'FETCH_SEARCH_TYPES.PENDING' };
      newState = reducer(oldState, action);
    });

    it('returns the correct state', () => {
      expect(newState).toMatchSnapshot();
    });
  });

  describe('when fetching search types succeeds', () => {
    beforeEach(() => {
      const action: FetchSearchTypesAction = {
        type: 'FETCH_SEARCH_TYPES.FULFILLED',
        payload: {
          'overnight-boarding': 'https://www.rover.com/api/v3/search/overnight-boarding/',
        },
      };
      newState = reducer(oldState, action);
    });

    it('returns the correct state', () => {
      expect(newState).toMatchSnapshot();
    });
  });

  describe('when fetching search types fails', () => {
    beforeEach(() => {
      const error = new Error('bang');
      const action: FetchSearchTypesAction = {
        type: 'FETCH_SEARCH_TYPES.REJECTED',
        payload: { error },
        error: true,
      };
      newState = reducer(oldState, action);
    });

    it('returns the correct state', () => {
      expect(newState).toMatchSnapshot();
    });
  });
});
