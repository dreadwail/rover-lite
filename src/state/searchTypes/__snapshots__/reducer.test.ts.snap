// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search types reducer when fetching search types fails returns the correct state 1`] = `
Object {
  "endpoints": Object {},
  "hasFailed": true,
  "isFetching": false,
}
`;

exports[`search types reducer when fetching search types starts returns the correct state 1`] = `
Object {
  "endpoints": Object {},
  "hasFailed": false,
  "isFetching": true,
}
`;

exports[`search types reducer when fetching search types succeeds returns the correct state 1`] = `
Object {
  "endpoints": Object {
    "overnight-boarding": "https://www.rover.com/api/v3/search/overnight-boarding/",
  },
  "hasFailed": false,
  "isFetching": false,
}
`;

exports[`search types reducer when passed no previous state returns the correct initial state 1`] = `
Object {
  "endpoints": Object {},
  "hasFailed": false,
  "isFetching": false,
}
`;
